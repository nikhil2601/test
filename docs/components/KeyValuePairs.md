[pep-comp](/) > [docs](/docs/README.md) > [components](/docs/components/README.md) > KeyValuePairs

--------------------------------------------------------------------------------

# KeyValuePairs

**src**: [KeyValuePairs.js](/src/lib/KeyValuePairs/KeyValuePairs.js)

`KeyValuePairs` renders a simple Key and Value pairs list.

## Example

```jsx
import React from 'react';
import { KeyValuePairs, Typography } from 'pep-comp';

const ITEMS = [
    {
        key: 'username',
        value: 'StarWars'
    },
    {
        key: 'email',
        value: 'luke@starwars.com'
    }
];

const KeyValueSample = () => (
    <div>
        <Typography type="title">Key Values Sample</Typography>
        <KeyValuePairs items={ITEMS} />
    </div>
);

export default KeyValueSample;
```

## Props

```javascript
/**
 * The `key` to extract from the item.
 * Key is inserted into the first column of the KeyValuePairs.
 */
itemKey    : PropTypes.string,
/**
 * The `value` to extract from the item.
 * Value is inserted into the second column of the KeyValuePairs.
 *
 * @type {[type]}
 */
itemValue  : PropTypes.string,
/**
 * The list of items for the KeyValuePairs
 */
items      : PropTypes.arrayOf(PropTypes.object).isRequired,
/**
 * The `onClick` prop for the item.
 */
onItemClick: PropTypes.func
```

## Default Props

```javascript
itemKey    : '',
onItemClick: null,
itemValue  : ''
```
