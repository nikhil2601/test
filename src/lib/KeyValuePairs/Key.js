import styled from 'styled-components';

import { getThemeProps } from 'utils/theme';

const KeyValueKeyStyled = styled.p`
    flex: 1 1 35%;
    text-align: right;
    /**
     * Add all of the remaining styles from theme
     */
    ${getThemeProps('KeyValuePairsKey.styles')};
`;

export default KeyValueKeyStyled;
