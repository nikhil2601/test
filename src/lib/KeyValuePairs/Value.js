import styled from 'styled-components';

import { getThemeProps } from 'utils/theme';

const KeyValueStyled = styled.p`
    flex: 1 1 65%;
    /**
     * Add all of the remaining styles from theme
     */
    ${getThemeProps('KeyValuePairsValue.styles')};
`;

export default KeyValueStyled;
