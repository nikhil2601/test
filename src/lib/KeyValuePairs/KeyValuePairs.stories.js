import React from 'react';
import { ThemeProvider } from 'styled-components';

import { THEME } from '../../constants/theme';
import Box from '../Box';
import KeyValuePairs from './KeyValuePairs';
import Typography from '../Typography';
import Switch from '../Switch';

export default {
    title: 'KeyValuePairs',
    component: KeyValuePairs,
};

const items = [
    {
        title: 'title1',
        value: 'value1',
    },
    {
        title: 'title2',
        value: 'value2',
    },
    {
        title: 'title3',
        value: 'value3',
    },
    {
        title: 'title4',
        value: 'value4',
    },
];

export const keyValuePairs = () => (
    <Box>
        <KeyValuePairs items={items} itemKey="title" itemValue="value" keySuffix=" :" />
    </Box>
);

const itemsWithCustomComponent = [
    {
        title: 'title1',
        value: 'value1',
    },
    {
        title: 'title2',
        value: 'value2',
    },
    {
        title: 'title3',
        value: <Switch checked />,
    },
    {
        title: 'title4',
        value: <Typography> Custom Value</Typography>,
    },
];

export const WithCustomValueComponent = () => (
    <ThemeProvider theme={THEME}>
        <Box>
            <KeyValuePairs
                items={itemsWithCustomComponent}
                itemKey="title"
                itemValue="value"
                keySuffix=" :"
            />
        </Box>
    </ThemeProvider>
);
