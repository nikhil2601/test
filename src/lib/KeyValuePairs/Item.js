import styled from 'styled-components';

import { getThemeProps } from 'utils/theme';

import Key from './Key';

const KeyItemStyled = styled.li`
    align-items: center;
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    width: 100%;
    ${Key} {
        margin-right: 10px;
    }
    /**
     * Add all of the remaining styles from theme
     */
    ${getThemeProps('KeyValuePairsItem.styes')};
`;

export default KeyItemStyled;
