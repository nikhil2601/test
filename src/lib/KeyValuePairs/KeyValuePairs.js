import PropTypes from 'prop-types';
import React from 'react';
import _get from 'lodash/get';
import _isFunction from 'lodash/isFunction';
import styled from 'styled-components';

import { genID } from 'utils/generate';
import { getThemeProps } from 'utils/theme';
import Divider from 'lib/Divider';

import Item from './Item';
import Key from './Key';
import Value from './Value';

const KeyValuePairsStyled = styled.ul`
    align-items: flex-start;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    list-style: none;
    margin: 0;
    padding: 0;
    /**
     * Add all of the remaining styles from theme
     */
    ${getThemeProps('KeyValuePairs.styles')};
`;

const ValueComponentStyled = styled.div`
    flex: 1 1 65%;
    ${getThemeProps('KeyValuePairsValue.styles')};
`;

const KeyValuePairs = ({ items, itemKey, onItemClick, itemValue, keySuffix, showDivider }) => (
    <KeyValuePairsStyled>
        {items.map(item => {
            const theKey = _get(item, itemKey, itemKey);
            const theValue = _get(item, itemValue, itemValue);
            const onClick = _isFunction(onItemClick) ? () => onItemClick(item) : null;
            const { [itemValue]: ValueComponent } = item;

            return (
                <React.Fragment key={genID()}>
                    <Item key={genID('KeyItem')} onClick={onClick}>
                        <Key>{`${theKey}${keySuffix}`}</Key>
                        {React.isValidElement(ValueComponent) ? (
                            <ValueComponentStyled>{ValueComponent}</ValueComponentStyled>
                        ) : (
                            <Value>{theValue}</Value>
                        )}
                    </Item>
                    {showDivider && <Divider />}
                </React.Fragment>
            );
        })}
    </KeyValuePairsStyled>
);

KeyValuePairs.propTypes = {
    /**
     * The `key` to extract from the item.
     * Key is inserted into the first column of the KeyValuePairs.
     */
    itemKey: PropTypes.string,
    /**
     * The `value` to extract from the item.
     * Value is inserted into the second column of the KeyValuePairs.
     *
     * @type {[type]}
     */
    itemValue: PropTypes.string,
    /**
     * The list of items for the KeyValuePairs
     */
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    keySuffix: PropTypes.string,
    /**
     * The `onClick` prop for the item.
     */
    onItemClick: PropTypes.func,

    showDivider: PropTypes.bool,
};

KeyValuePairs.defaultProps = {
    itemKey: '',
    itemValue: '',
    keySuffix: '',
    onItemClick: null,
    showDivider: false,
};

export default KeyValuePairs;
