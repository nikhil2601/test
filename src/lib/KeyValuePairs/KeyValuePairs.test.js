import React from 'react';
import { shallow } from 'enzyme';

import KeyValuePairs from './KeyValuePairs';

describe('KeyValuePairs', () => {
    it('should render correctly in "debug" mode', () => {
        const component = shallow(<KeyValuePairs items={[]} debug />);

        expect(component).toMatchSnapshot();
    });
});
