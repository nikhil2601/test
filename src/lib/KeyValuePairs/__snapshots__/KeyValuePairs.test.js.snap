// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyValuePairs should render correctly in "debug" mode 1`] = `<KeyValuePairs__KeyValuePairsStyled />`;
